File: api/Pin.go
package api

import (
	"github.com/lvcdy/huawei-lte-api-go/session"
)

// Pin 对应 Pin.py。
type Pin struct {
	*session.ApiGroup
}

// NewPin 创建 Pin API 分组。
func NewPin(s *session.Session) *Pin {
	return &Pin{ApiGroup: session.NewApiGroup(s)}
}

// Status 对应 pin/status。
func (p *Pin) Status() (map[string]interface{}, error) {
	return p.S.Get("pin/status", nil, "api")
}

// Simlock 对应 pin/simlock。
func (p *Pin) Simlock() (map[string]interface{}, error) {
	return p.S.Get("pin/simlock", nil, "api")
}

// SavePin 对应 pin/save-pin。
func (p *Pin) SavePin() (map[string]interface{}, error) {
	return p.S.Get("pin/save-pin", nil, "api")
}

// Operate 执行 PIN 操作。对应 pin/operate (post_set, is_encrypted)。
//
// operate_type: 0 - 验证 PIN, 1 - 启用 PIN 验证, 2 - 禁用 PIN 验证,
// 3 - 设置新 PIN, 4 - 使用 PUK 码。
func (p *Pin) Operate(operateType string, currentPin *string, newPin *string, pukCode *string) (interface{}, error) {
	data := map[string]interface{}{
		"OperateType": operateType,
		"CurrentPin":  nil,
		"NewPin":      nil,
		"PukCode":     nil,
	}
	if currentPin != nil {
		data["CurrentPin"] = *currentPin
	}
	if newPin != nil {
		data["NewPin"] = *newPin
	}
	if pukCode != nil {
		data["PukCode"] = *pukCode
	}
	return p.S.PostSet("pin/operate", data, false, "api", true, false)
}

File: src/api/pin.rs
//! Pin API group (`api/Pin.py`).

use serde_json::Value;

use crate::connection::Connection;
use crate::errors::Result;
use crate::tools::map_of;
use crate::xml::XmlMap;

use super::{get_value, post_set};

/// Pin API group.
pub struct Pin<'a> {
    conn: &'a Connection,
}

impl<'a> Pin<'a> {
    /// Build a group bound to `conn`.
    pub fn new(conn: &'a Connection) -> Self {
        Pin { conn }
    }

    /// `pin/status`. PIN status.
    pub fn status(&self) -> Result<Value> {
        get_value(self.conn, "api/pin/status")
    }

    /// `pin/simlock`. SIM lock status.
    pub fn simlock(&self) -> Result<Value> {
        get_value(self.conn, "api/pin/simlock")
    }

    /// `pin/save-pin`. Save PIN.
    pub fn save_pin(&self) -> Result<Value> {
        get_value(self.conn, "api/pin/save-pin")
    }

    /// Perform an operation on the PIN.
    ///
    /// * `operate_type` — `0` verify, `1` enable, `2` disable, `3` set new,
    ///   `4` use PUK.
    /// * `current_pin` / `new_pin` / `puk_code` — optional, empty when unused.
    pub fn operate(
        &self,
        operate_type: &str,
        current_pin: Option<&str>,
        new_pin: Option<&str>,
        puk_code: Option<&str>,
    ) -> Result<String> {
        let body: XmlMap = map_of([
            ("OperateType", operate_type.to_string()),
            ("CurrentPin", current_pin.unwrap_or("").to_string()),
            ("NewPin", new_pin.unwrap_or("").to_string()),
            ("PukCode", puk_code.unwrap_or("").to_string()),
        ]);
        post_set(self.conn, "api/pin/operate", &body)
    }
}
